AWS and GCP configurations for qe-sap-deployment

mpagot opened the pull request:
Add two config.yaml for the qesapdeploy test.

Bump config apiver to 2 for all qesap and trento configurations

qesapdeploy lib improvements:
 - remove obsolete functions using build.sh
 - fix bug in qesap_upload_logs about multiple uploads of same logs
 - fix log file name for qesap_execute
 - croak for qesap_ansible_cmd
 - add unit testing about qesap_execute

Adapt Trento lib and qesap tests to the new qesap_execute return mode cmr_mon from hacluster in qesap test

Verification run:
 - [AZURE_RUN_1](http://openqaworker15.qa.suse.cz/tests/51482) [AZURE_RUN_2](http://openqaworker15.qa.suse.cz/tests/51603).....OK 
 - [GCP_RUN_1](http://openqaworker15.qa.suse.cz/tests/51475) [GCP_RUN_2](http://openqaworker15.qa.suse.cz/tests/51604).....OK
 - [AWS_RUN_1](http://openqaworker15.qa.suse.cz/tests/51540)  [AWS_RUN_2](http://openqaworker15.qa.suse.cz/tests/51602)......CONFIGURE and TERRAFORM OK

mpagot commented on data/sles4sap/qe_sap_deployment/trento_azure.yaml:
```diff
@@ -1,5 +1,5 @@
 provider: "%PROVIDER%"
-apiver: 1
+apiver: 2
```
This one is about (but does not needs) https://github.com/SUSE/qe-sap-deployment/pull/79

lpalovsky commented on lib/qesapdeployment.pm:
```diff
@@ -124,11 +120,9 @@ sub qesap_upload_logs {
     my $failok = $args{failok};
     record_info("Uploading logfiles", join("\n", @log_files));
-    for my $file (@log_files) {
+    while (my $file = pop @log_files) {
```
I like this. This is cool :)

lpalovsky reviewed (approved):
LGTM

alvarocarvajald commented on data/sles4sap/qe_sap_deployment/qesap_aws.yaml:
```diff
@@ -0,0 +1,35 @@
+    aws_region: "%REGION%"
+    deployment_name: "%DEPLOYMENTNAME%"
+    os_image: "%QESAP_CLUSTER_OS_VER%"
+    os_owner: "amazon"
```
This needs to be a `%VARIABLE%` also. On QAM we would test with ami's owned by amazon, while on Product QA, we would test with a self owned image uploaded by an openQA job.

Keep in mind that my comment is moot if we have a `qesap_aws.yaml` file for QAM and another for Product QA.

mpagot replied:
`qesap_aws.yaml` can be considered dedicated to this test only.
I'm able to find other files using the os_owner variable in the repo, but all of them has a value hard-coded to "amazon" or "self". Is there any example of a file using a template variable for it. I'm mostly interested to check where the template gets the value from.

alvarocarvajald replied:
Yes. Here: http://mango.qa.suse.de/tests/5023#step/hana_sr_deployment/71

It's getting the value from the PUBLIC_CLOUD_IMAGE_OWNER setting.

More details in my branch from 2 weeks ago:

- https://github.com/alvarocarvajald/os-autoinst-distri-opensuse/commit/8ee61157dd4bee861f05add0ced8f6966afd6e35#diff-3226f0b86bf0b058a92784f787d19467ebecc5246fbcc5e7750455d3f63734f6R12
- https://github.com/alvarocarvajald/os-autoinst-distri-opensuse/commit/8ee61157dd4bee861f05add0ced8f6966afd6e35#diff-98a44adfecdae80a9b5eab7a2922a193b903d8fa846b355db78783ff85c13b9bR45

mpagot replied:
ok, I was looking that on https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/master/variables.md

I plan only to quse these qesap_*.yaml for the tests/sles4sap/qesapdeploy/* tests and accept the limitation only to run those tests agains amazon owned images. Does it make sense?

alvarocarvajald replied:
Yes, it does.

alvarocarvajald commented on lib/qesapdeployment.pm:
```diff
@@ -356,7 +208,9 @@ sub qesap_execute {
     my %paths = qesap_get_file_paths();
-    my $exec_log = "/tmp/qesap_exec_$args{cmd}_$args{cmd_options}.log.txt";
+    my $log_postfix = $args{cmd_options} ? "_$args{cmd_options}" : "";
+    my $exec_log = "/tmp/qesap_exec_$args{cmd}$log_postfix.log.txt";
```
Don't see the need for `$log_postfix` variable. You can accomplish the same by:
```
my $exec_log = $args{cmd_options} ? "/tmp/qesap_exec_$args{cmd}_$args{cmd_options}.txt" : "/tmp/qesap_exec_$args{cmd}.txt";
```
If you feel that line is too long and repeating too much text, you could also do:
```
my $exec_log = "/tmp/qesap_exec_$args{cmd}";
$exec_log .= "_$args{cmd_options}" if ($args{cmd_options});
$exec_log .= '.txt';
```

alvarocarvajald commented on lib/qesapdeployment.pm:
```diff
@@ -356,7 +208,9 @@ sub qesap_execute {
-    my $exec_log = "/tmp/qesap_exec_$args{cmd}_$args{cmd_options}.log.txt";
+    my $log_postfix = $args{cmd_options} ? "_$args{cmd_options}" : "";
+    my $exec_log = "/tmp/qesap_exec_$args{cmd}$log_postfix.log.txt";
+    $args{cmd_options} ||= '';
```
I guess this line should go above L211

alvarocarvajald commented on lib/trento.pm:
```diff
@@ -286,10 +286,11 @@ Deploy a SAP Landscape using a previously configured qe-sap-deployment
-    qesap_execute(cmd => 'terraform', verbose => 1, timeout => 1800);
-    qesap_execute(cmd => 'ansible', verbose => 1, timeout => 1800);
+    my $ret = qesap_execute(cmd => 'terraform', verbose => 1, timeout => 1800);
+    die "'qesap.py terraform' return: $ret" unless $ret == 0;
```
I'd go with `if ($ret)` (`$ret` equal to zero will evaluate to false)

alvarocarvajald commented on lib/trento.pm:
```diff
@@ -286,10 +286,11 @@ Deploy a SAP Landscape using a previously configured qe-sap-deployment
+    my $ret = qesap_execute(cmd => 'terraform', verbose => 1, timeout => 1800);
+    die "'qesap.py terraform' return: $ret" unless $ret == 0;
+    $ret = qesap_execute(cmd => 'ansible', verbose => 1, timeout => 1800);
+    die "'qesap.py ansible' return: $ret" unless $ret == 0;
```
Same as above. I'd go with `if ($ret)`

alvarocarvajald commented on lib/trento.pm:
```diff
@@ -299,8 +300,10 @@ Destroy the qe-sap-deployment SAP Landscape
-    qesap_execute(cmd => 'ansible', cmd_options => '-d', verbose => 1, timeout => 300);
-    qesap_execute(cmd => 'terraform', cmd_options => '-d', verbose => 1, timeout => 1200);
+    my $ret = qesap_execute(cmd => 'ansible', cmd_options => '-d', verbose => 1, timeout => 300);
+    die "'qesap.py ansible -d' return: $ret" unless $ret == 0;
```
Same as above. I'd go with `if ($ret)`

alvarocarvajald commented on lib/trento.pm:
```diff
@@ -299,8 +300,10 @@ Destroy the qe-sap-deployment SAP Landscape
+    my $ret = qesap_execute(cmd => 'ansible', cmd_options => '-d', verbose => 1, timeout => 300);
+    die "'qesap.py ansible -d' return: $ret" unless $ret == 0;
+    $ret = qesap_execute(cmd => 'terraform', cmd_options => '-d', verbose => 1, timeout => 300);
+    die "'qesap.py terraform -d' return: $ret" unless $ret == 0;
```
Same as above. I'd go with `if ($ret)`

alvarocarvajald commented on tests/sles4sap/qesapdeployment/deploy.pm:
```diff
@@ -11,8 +11,10 @@ use testapi;
-    qesap_execute(cmd => 'terraform', verbose => 1, timeout => 1800);
-    qesap_execute(cmd => 'ansible', verbose => 1, timeout => 1800);
+    my $ret = qesap_execute(cmd => 'terraform', verbose => 1, timeout => 1800);
+    die "'qesap.py terraform' return: $ret" unless $ret == 0;
```
Same as above. I'd go with `if ($ret)`

alvarocarvajald commented on tests/sles4sap/qesapdeployment/deploy.pm:
```diff
@@ -11,8 +11,10 @@ use testapi;
+    my $ret = qesap_execute(cmd => 'terraform', verbose => 1, timeout => 1800);
+    die "'qesap.py terraform' return: $ret" unless $ret == 0;
+    $ret = qesap_execute(cmd => 'ansible', verbose => 1, timeout => 1800);
+    die "'qesap.py ansible' return: $ret" unless $ret == 0;
```
Same as above. I'd go with `if ($ret)`

alvarocarvajald commented on tests/sles4sap/qesapdeployment/destroy.pm:
```diff
@@ -14,8 +14,10 @@ sub run {
-    qesap_execute(cmd => 'ansible', cmd_options => '-d', verbose => 1, timeout => 300);
-    qesap_execute(cmd => 'terraform', cmd_options => '-d', verbose => 1, timeout => 1200);
+    my $ret = qesap_execute(cmd => 'ansible', cmd_options => '-d', verbose => 1, timeout => 300);
+    die "'qesap.py ansible -d' return: $ret" unless $ret == 0;
```
Same as above. I'd go with `if ($ret)`

alvarocarvajald commented on tests/sles4sap/qesapdeployment/destroy.pm:
```diff
@@ -14,8 +14,10 @@ sub run {
+    my $ret = qesap_execute(cmd => 'ansible', cmd_options => '-d', verbose => 1, timeout => 300);
+    die "'qesap.py ansible -d' return: $ret" unless $ret == 0;
+    $ret = qesap_execute(cmd => 'terraform', cmd_options => '-d', verbose => 1, timeout => 900);
+    die "'qesap.py terraform -d' return: $ret" unless $ret == 0;
```
Same as above. I'd go with `if ($ret)`

alvarocarvajald commented on tests/sles4sap/qesapdeployment/test_cluster.pm:
```diff
@@ -19,7 +18,7 @@ sub run {
     qesap_ansible_cmd(cmd => $_, provider => $prov, $_) for ('pwd', 'uname -a', 'cat /etc/os-release');
     qesap_ansible_cmd(cmd => 'ls -lai /hana/', provider => $prov, filter => 'hana');
-    qesap_ansible_cmd(cmd => $_, provider => $prov, filter => 'vmhana01'), for ('crm status', 'crm_mon -R -r -n -N -1');
+    qesap_ansible_cmd(cmd => $_, provider => $prov, filter => 'vmhana01'), for ('crm status', $crm_mon_cmd);
```
Nice! Thank you!

mpagot replied:
This code is fine thanks to `use hacluster '$crm_mon_cmd';`

Using a more generic `use hacluster;` result in an error about a double definition of post_fail_hook

alvarocarvajald replied:
Makes sense. You could've probably also used `$hacluster::crm_mon_cmd`.

alvarocarvajald reviewed (commented):
There are a lot of functions removed from `lib/qesapdeployment.pm` ... are these not necessary anymore or are you simply relocating them to another library?

I added some more comments inline.

mpagot commented:
> There are a lot of functions removed from `lib/qesapdeployment.pm` ... are these not necessary anymore or are you simply relocating them to another library?
> 
> I added some more comments inline.

I removed all the functions that are about build.sh old glue script

alvarocarvajald reviewed (approved):
LGTM
